### ISSUE TITLE
Unable to attach with the SafeNativeAgentLoader

### ISSUE THREAD

**Issue by ttiimm** [vsilaev/tascalate-instrument#2]
I was trying to use the Agent attach feature with a version of Temurin JRE (11.0.17), but ran across an issue.

The AgentLoaders tries to attach using the SafeNativeAgentLoader, however this doesn't appear to work. Here is an excerpt from the failing stacktrace.

```
  exc: java.lang.UnsatisfiedLinkError: Error looking up function 'Agent_OnAttach': /opt/.../jdk-11.0.17+8-jre//bin/java: undefined symbol: Agent_OnAttach
  exc: 	at com.sun.jna.Function.<init>(Function.java:252)
  exc: 	at com.sun.jna.NativeLibrary.getFunction(NativeLibrary.java:620)
  exc: 	at com.sun.jna.NativeLibrary.getFunction(NativeLibrary.java:596)
  exc: 	at com.sun.jna.NativeLibrary.getFunction(NativeLibrary.java:582)
  exc: 	at com.sun.jna.Library$Handler.invoke(Library.java:248)
  exc: 	at net.tascalate.instrument.attach.core.$Proxy123.Agent_OnAttach(Unknown Source)
  exc: 	at net.tascalate.instrument.attach.core.NativeAgentLoader.attach(Unknown Source)
  exc: 	at net.tascalate.instrument.attach.core.SafeNativeAgentLoader.attach(Unknown Source)
  exc: 	at net.tascalate.instrument.attach.api.AgentLoaders$WithArgsCheck.attach(Unknown Source)
  exc: 	at net.tascalate.instrument.attach.api.AgentLoaders.attach(Unknown Source)
```

The same code is working with the Temurin JDK (11.0.18).

Do you have any insight what might be failing?

**Comment by vsilaev** [comment]
First, native agent loader will work correctly only with JDK while JRE has no necessary functionality exposed.
However, what you reported is indeed an error with SafeNativeAgentLoader -- it should catch this error and report back native agent loader as unavailable.
Will fix the issue in the next release.

**Comment by ttiimm** [comment]
Cool -- that makes sense.

Do you expect either the LocalAgentLoader or ExternalAgentLoader to work with a JRE? 

Is there a way to ask the chain to skip over the SafeNativeAgentLoader?

**Comment by vsilaev** [comment]
The only option to bypass `SafeNativeAgentLoader `is to exclude JNA from class path as of now. And you should use explicitly `LocalAgentLoader` (by default it downgrades to `ExternalAgentLoader` if necessary)

> Do you expect either the LocalAgentLoader or ExternalAgentLoader to work with a JRE?

Unfortunately, since Java 11 the `jdk.attach` module is not included in JRE by default. So you either should use JDK or create your custom JRE with `jlink` command (obviously, incl. `jdk.attch`). More info is [here](https://support.azul.com/hc/en-us/articles/13078370139412-Creating-a-JRE-from-the-JDK-with-jlink)

Btw, I've released version 1.3.3 with the fix for the issue

**Comment by vsilaev** [comment]
It seems that unlike Oracle JDK-s Temurin  link `instrument` dll dynamically rather than statically...

**Comment by vsilaev** [comment]
Released a version [1.4.0](https://github.com/vsilaev/tascalate-instrument/releases/tag/1.4.0) - now JNA should work with JRE and JDK may work with either `jdk.attach` or `com.sun.jna`. Please read README for the updated info.
